presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java

```diff
@@ -795,15 +799,86 @@ public PrestoS3InputStream(AmazonS3 s3, String host, Path path, int maxAttempts,
         @Override
         public void close()
         {
-            closed = true;
+            closed.set(true);
             closeStream();
         }
 
+        @Override
+        public int read(long position, byte[] buffer, int offset, int length)
+                throws IOException
+        {
+            checkClosed();
+            if (position < 0) {
+                throw new EOFException(FSExceptionMessages.NEGATIVE_SEEK);
+            }
+            checkPositionIndexes(offset, offset + length, buffer.length);
+            if (length == 0) {
+                return 0;
+            }
+
+            try {
+                return retry()
+                        .maxAttempts(maxAttempts)
+                        .exponentialBackoff(BACKOFF_MIN_SLEEP, maxBackoffTime, maxRetryTime, 2.0)
+                        .stopOn(InterruptedException.class, UnrecoverableS3OperationException.class, EOFException.class)
+                        .onRetry(STATS::newGetObjectRetry)
+                        .run("getS3Object", () -> {
+                            InputStream in;
+                            try {
+                                GetObjectRequest request = new GetObjectRequest(host, keyFromPath(path))
+                                        .withRange(position, (position + length) - 1);
+                                in = s3.getObject(request).getObjectContent();
+                            }
+                            catch (RuntimeException e) {
+                                STATS.newGetObjectError();
+                                if (e instanceof AmazonS3Exception) {
+                                    switch (((AmazonS3Exception) e).getStatusCode()) {
+                                        case HTTP_RANGE_NOT_SATISFIABLE:
+                                            throw new EOFException(FSExceptionMessages.CANNOT_SEEK_PAST_EOF);
```

Comment:
Does it throw `HTTP_RANGE_NOT_SATISFIABLE` if range exceeds file length but there are some bytes to be read?

Reply:
I would assume so, although their documentation doesn't say. It won't matter for us since we always use it with valid ranges.